README
A barely playable carrom game. Still a work in progress.


NEWS
NEWS file for gnarrom

src/main.cpp
#include <SFML/System.hpp>
#include "game.hpp"
#include <gtkmm.h>

//Create a separate global thread for our game, so that the gtk frontend and the game
//run in different threads

class launcher
{
private:
    Gtk::Window &window;
    Gtk::SpinButton &winsize_spin;
    static sf::Thread game;
public:
    void startgame();
    launcher(Gtk::Window &window, Gtk::SpinButton &winsize_spin) : window(window), winsize_spin(winsize_spin){};
};

sf::Thread launcher::game(&game::init);

void launcher::startgame()
{
    window.iconify();
    int winsize_int = winsize_spin.get_value_as_int();
    game::setwindowsize(winsize_int);
    game.launch();
}

int main(int argc, char **argv) {
    Gtk::Main kit(argc, argv);

    Gtk::Window window;

    Gtk::SpinButton winsize_spin;
	
    launcher gnarrom(window, winsize_spin);

    Gtk::VBox mainbox;
    Gtk::HBox controlbox;
    Gtk::HBox winsizebox;
    
    Gtk::Button start("Launch");
    start.show();
    start.signal_clicked().connect(sigc::mem_fun(&gnarrom, &launcher::startgame));

    Gtk::Button stop("Quit");
    stop.show();
    stop.signal_clicked().connect(sigc::ptr_fun(&Gtk::Main::quit));
    
    Gtk::Label winsizelabel("Window Size");
    winsizelabel.show();

    winsize_spin.set_range(400, 1200);
    winsize_spin.set_increments(1, 10);
    winsize_spin.set_value(700);
    winsize_spin.show();

    controlbox.pack_start(start);
    controlbox.pack_end(stop);
    controlbox.show();
    
    winsizebox.pack_start(winsizelabel);
    winsizebox.pack_end(winsize_spin);
    winsizebox.show();

    mainbox.pack_start(winsizebox);
    mainbox.pack_end(controlbox);
    mainbox.show();
    
    window.add(mainbox);

    Gtk::Main::run(window);
    
    return 0;
}


includes/game.hpp
#ifndef GAME_H
#define GAME_H

#include <SFML/Graphics.hpp>
#include <SFML/Audio.hpp>
#include "ingameobjects.hpp"

class objectlist;

class game
{
private:
    static sf::RenderWindow window;
    enum gameStates {_init, _new, _striker, _pointer, _hit, _motion,  _quit};
    static gameStates _gamestate;
    static sf::Texture brd;
    static sf::Texture cointex;
    static sf::Texture strikertex;
    static board B;
    static void loop();
    static void changeState(sf::Event &event);
    static void captureMouse();
    static void close();
    static int player_count;
    static int player;
    static int turn;
    static int turnincr;
    static int coinsinpocket;
    static int windowsize;
    static bool capmouse;
    static sf::SoundBuffer soundbuf;
    static sf::Sound sound;
    static sf::Time frametime;
    static sf::Color brdcolor;
    static sf::Color wcolor;
    static sf::Color bcolor;
    static sf::Color qcolor;
    static sf::Color scolor;
    static sf::Color pointercolorin;
    static sf::Color pointercolorout;
    static objectlist *objects;
    static void processPocket(int i);
    static void processFine();
    static void endturn();
public:
    static void init();
    static void playsound(float impact);
    static void setwindowsize(int winsize);
    static void setColor(int brdR,int brdG,int brdB,
			 int wR, int wG, int wB,
			 int bR, int bG, int bB,
			 int qR, int qG, int qB,
			 int sR, int sG, int sB,
			 int piR, int piG, int piB,
			 int poR, int poG, int poB);
    friend class board;
    friend class gameObject;
    friend class striker;
    friend class coin;
    friend class pointer;
    friend class objectlist;
    friend void collision(gameObject* A, gameObject* B);
    friend void collisionloop();
};

#endif


src/ingameobjects.cpp
#include <SFML/System.hpp>
#include <SFML/Window.hpp>
#include <SFML/Graphics.hpp>
#include <math.h>
#include "gameobjectmanager.hpp"
#include "game.hpp"
#include "ingameobjects.hpp"

striker::striker()
{
    circle.setRadius(0.0267f*game::windowsize);
    circle.setOrigin(circle.getRadius(), circle.getRadius());
    circle.setPosition(game::windowsize/2.0f, game::windowsize/2.0f);
    circle.setFillColor(game::scolor);
    circle.setOutlineThickness(-3.f);
    circle.setOutlineColor(game::scolor);
    circle.setTexture(&game::strikertex);
    sf::Mouse::setPosition(sf::Vector2i(300, 495), game::window);
    relMousePos = sf::Vector2f(300, 495);
    vel_x = vel_init;
    vel_y = vel_init;
    massfactor = 2.8f;
    restitution = 0.9f;
    frictionfactor = 1.f;
    position = circle.getPosition();
}

void striker::update()
{
    position = circle.getPosition();
    if (game::_gamestate == game::_striker) {
        input();
        boundtoplay();
    }
    else if (game::_gamestate == game::_motion) {
        move();
    }
    boundtoboard();
    game::window.draw(circle);
}

void striker::input()
{
    if (sf::Joystick::isConnected(game::player)) {
        circle.move(0.03f*sf::Joystick::getAxisPosition(game::player, sf::Joystick::X),
                0.03f*sf::Joystick::getAxisPosition(game::player, sf::Joystick::Y));
    }
    else {
        sf::Vector2i mousePos = sf::Mouse::getPosition(game::window);
        sf::Vector2f circleMove(mousePos.x - relMousePos.x, mousePos.y - relMousePos.y);
        circle.setPosition(circle.getPosition().x + circleMove.x, circle.getPosition().y + circleMove.y);
        game::captureMouse();
        relMousePos = sf::Vector2f(sf::Mouse::getPosition(game::window).x, sf::Mouse::getPosition(game::window).y);
    }
}

void striker::boundtoplay(){

    for (int i = 1; i < 20; i++) {

        sf::Vector2f posA = circle.getPosition();
        sf::Vector2f posB = game::objects->list[i]->circle.getPosition();

        float dist = sqrt(pow(posA.x - posB.x, 2) + pow(posA.y - posB.y, 2));

        if (dist < circle.getRadius() + game::objects->list[i]->circle.getRadius() + 0.001f) {
            if (posA.x < posB.x) {
                circle.setPosition(posB.x + circle.getRadius() +
                        game::objects->list[i]->circle.getRadius() + 0.0015f,
                        posA.y);
            }
            else {
                circle.setPosition(posB.x - circle.getRadius() -
                        game::objects->list[i]->circle.getRadius() - 0.0015f,
                        posA.y);
            }
        }

        posA = circle.getPosition();
        dist = sqrt(pow(posA.x - posB.x, 2) + pow(posA.y - posB.y, 2));

        if (dist < circle.getRadius() + game::objects->list[i]->circle.getRadius() + 0.001f) {
            circle.setPosition(game::windowsize/2.0f, posA.y);
        }
    }

    sf::Vector2f circlePos = circle.getPosition();
    float x = circlePos.x;
    float y = circlePos.y;
    if (game::player == 0) {
        if (y < board::bottom - board::playzonein ) {
            circle.setPosition(sf::Vector2f(x, board::bottom - board::playzonein));
        }
        if (y > board::bottom - board::playzoneout) {
            circle.setPosition(sf::Vector2f(x, board::bottom - board::playzoneout));
        }

        circlePos = circle.getPosition();
        x = circlePos.x;
        y = circlePos.y;

        if(x + y < game::windowsize + circle.getRadius() + 0.011f*game::windowsize) {
            circle.setPosition(game::windowsize + circle.getRadius() + 0.011f*game::windowsize - y, y);
        }

        else if(y - x < circle.getRadius() + 0.011f*game::windowsize) {
            circle.setPosition(y - circle.getRadius() - 0.011f*game::windowsize, y);
        }
    }

    else if (game::player == 1) {
        if (y > board::top + board::playzonein ) {
            circle.setPosition(sf::Vector2f(x, board::top + board::playzonein));
        }
        if (y < board::top + board::playzoneout) {
            circle.setPosition(sf::Vector2f(x, board::top + board::playzoneout));
        }

        circlePos = circle.getPosition();
        x = circlePos.x;
        y = circlePos.y;

        if(x + y > game::windowsize - circle.getRadius() - 0.011f*game::windowsize) {
            circle.setPosition(game::windowsize - circle.getRadius() - 0.011f*game::windowsize - y, y);
        }

        else if(y - x + circle.getRadius() + 0.011f*game::windowsize > 0) {
            circle.setPosition(y + circle.getRadius() + 0.011f*game::windowsize, y);
        }
    }

    for (int i = 1; i < 20; i++) {

        sf::Vector2f posA = circle.getPosition();
        sf::Vector2f posB = game::objects->list[i]->circle.getPosition();

        float dist = sqrt(pow(posA.x - posB.x, 2) + pow(posA.y - posB.y, 2));

        if (dist < circle.getRadius() + game::objects->list[i]->circle.getRadius() + 0.001f) {
            if (posA.x < game::windowsize/2.0) {
                circle.setPosition(posB.x + circle.getRadius() +
                        game::objects->list[i]->circle.getRadius() + 0.0015f , posA.y);
            }
            else {
                circle.setPosition(posB.x - circle.getRadius() -
                        game::objects->list[i]->circle.getRadius() - 0.0015f, posA.y);
            }
        }
    }

}

void striker::reset()
{
    vel_x = vel_init;
    vel_y = vel_init;
    circle.setPosition(game::windowsize/2.0, game::windowsize/2.0);
    game::_gamestate = game::_striker;
    game::objects->setActive(0);
}

coin::coin(char t)
{
    circle.setRadius(0.019f*game::windowsize);
    circle.setOrigin(circle.getRadius(), circle.getRadius());
    circle.setOutlineThickness(-0.005*game::windowsize);
    circle.setPosition(300, 300);
    circle.setTexture(&game::cointex);
    type = t;
    switch(type){
        case 'w':
            circle.setFillColor(game::wcolor);
            circle.setOutlineColor(game::wcolor);
            break;
        case 'b':
            circle.setFillColor(game::bcolor);
            circle.setOutlineColor(game::bcolor);
            break;
        default:
            circle.setFillColor(game::qcolor);
            circle.setOutlineColor(game::qcolor);
    }
    vel_x = vel_init;
    vel_y = vel_init;
    massfactor = 1.f;
    restitution = 0.95f;
    frictionfactor = 3.f;
}

void coin::update()
{
    if (game::_gamestate == game::_motion) {
        move();
    }
    boundtoboard();
    game::window.draw(circle);
}

void coin::reset()
{
    vel_x = vel_init;
    vel_y = vel_init;
    circle.setPosition(game::windowsize/2.0, game::windowsize/2.0);

    for (int i = 1; i < 20; i++) {
        if (this != game::objects->list[i]) {
            sf::Vector2f posA = circle.getPosition();
            sf::Vector2f posB = game::objects->list[i]->circle.getPosition();

            float dist = sqrt(pow(posA.x - posB.x, 2) + pow(posA.y - posB.y, 2));

            if (dist < circle.getRadius() + game::objects->list[i]->circle.getRadius() + 0.001f) {
                if (posA.x > posB.x) {
                    circle.setPosition(posB.x + circle.getRadius() +
                            game::objects->list[i]->circle.getRadius() + 0.0015f , posA.y);
                }
                else {
                    circle.setPosition(posB.x - circle.getRadius() -
                            game::objects->list[i]->circle.getRadius() - 0.0015f, posA.y);
                }
            }
        }
    }
}

void board::setTexture()
{
    this->brd.setTexture(game::brd);
    brd.setScale(game::windowsize/2268.f, game::windowsize/2268.f);
    top = floor(0.033f*game::windowsize);
    bottom = game::windowsize - top;
    left = top;
    right = bottom;
    board::playzonein = floor(0.143*game::windowsize);
    board::playzoneout = floor(0.135*game::windowsize);
    board::pocket_offset = playzoneout;
}

void board::draw()
{
    game::window.draw(brd);
}


pointer::pointer()
{
    circle.setRadius(3.f);
    circle.setOrigin(3.f, 3.f);
    circle.setOutlineThickness(-1.f);
    circle.setFillColor(game::pointercolorin);
    circle.setOutlineColor(game::pointercolorout);
    relMousePos = sf::Vector2f(sf::Mouse::getPosition(game::window).x, sf::Mouse::getPosition(game::window).y);
    vel_x = 0.f;
    vel_y = 0.f;
    massfactor = 2.5f;
    restitution = 1.f;
    frictionfactor = 0.f;
}

void pointer::update()
{
    static float dist = 0.f;
    if(game::_gamestate == game::_striker){
        circle.setPosition(striker::position);
        relMousePos = sf::Vector2f(sf::Mouse::getPosition(game::window).x, sf::Mouse::getPosition(game::window).y);
        return;
    }

    else if (game::_gamestate == game::_pointer) {
        input();
        boundaroundstriker();
        dist = sqrt(pow(circle.getPosition().x - striker::position.x,2) +
                pow(circle.getPosition().y - striker::position.y,2));
    }
    else if (game::_gamestate == game::_hit) {
        hitStriker(dist);
    }
    game::window.draw(circle);
}

void pointer::input()
{
    if (sf::Joystick::isConnected(game::player)) {
        circle.move(0.02f*sf::Joystick::getAxisPosition(game::player, sf::Joystick::X),
                0.02f*sf::Joystick::getAxisPosition(game::player, sf::Joystick::Y));
    }
    else {
        sf::Vector2i mousePos = sf::Mouse::getPosition(game::window);
        sf::Vector2f circleMove(mousePos.x - relMousePos.x, mousePos.y - relMousePos.y);
        circle.setPosition(circle.getPosition().x + circleMove.x, circle.getPosition().y + circleMove.y);
        game::captureMouse();
        relMousePos = sf::Vector2f(sf::Mouse::getPosition(game::window).x, sf::Mouse::getPosition(game::window).y);
    }
}


void pointer::boundaroundstriker()
{
    float x = circle.getPosition().x - striker::position.x;
    float y = circle.getPosition().y - striker::position.y;
    float o = atan2f(y, x);
    float r = sqrt(x*x + y*y);

    if (r > 0.135f*game::windowsize) {
        r = 0.135*game::windowsize;
    }
    else if (r < 0.03*game::windowsize){
        r = 0.03*game::windowsize;
    }

    x = r*cos(o);
    y = r*sin(o);

    circle.setPosition(striker::position.x + x, striker::position.y + y);
}

void pointer::hitStriker(float dist)
{
    float x = circle.getPosition().x - striker::position.x;
    float y = circle.getPosition().y - striker::position.y;
    float o = atan2(y, x);
    vel_x = -15.5f*dist/(0.03*sqrt(game::windowsize))*cos(o);
    vel_y = -15.5f*dist/(0.03*sqrt(game::windowsize))*sin(o);
    circle.move(vel_x*game::frametime.asSeconds(), vel_y*game::frametime.asSeconds());
}


void gameObject::boundtoboard()
{
    float impact = sqrt(vel_x*vel_x + vel_y*vel_y);
    sf::Vector2f circlePos = circle.getPosition();
    if (circlePos.x - circle.getRadius() < board::left ) {
        circle.setPosition(sf::Vector2f(board::left + circle.getRadius(), circlePos.y));
        vel_x *= -board::edgerestitution;
	game::playsound(impact);
    }
    if (circlePos.x + circle.getRadius() > board::right ) {
        circle.setPosition(sf::Vector2f(board::right - circle.getRadius(), circlePos.y));
        vel_x *= -board::edgerestitution;
	game::playsound(impact);
    }
    if (circlePos.y - circle.getRadius() < board::top ) {
        circle.setPosition(sf::Vector2f(circlePos.x, board::top + circle.getRadius()));
        vel_y *= -board::edgerestitution;
	game::playsound(impact);
    }
    if (circlePos.y + circle.getRadius() > board::bottom ) {
        circle.setPosition(sf::Vector2f(circlePos.x, board::bottom - circle.getRadius()));
        vel_y *= -board::edgerestitution;
	game::playsound(impact);
    }
}

void gameObject::move()
{

    if (fabs(vel_x) < 10.f)
        vel_x = vel_init;
    if (fabs(vel_y) < 10.f)
        vel_y = vel_init;

    circle.move(vel_x*game::frametime.asSeconds(), 0);
    circle.move(0, vel_y*game::frametime.asSeconds());


    if (fabs(vel_x) > vel_init)
        vel_x -= board::frictioncoefficient*massfactor*frictionfactor*game::frametime.asSeconds()*(vel_x)/(sqrt(vel_x*vel_x + vel_y*vel_y));
    if (fabs(vel_y) > vel_init)
        vel_y -= board::frictioncoefficient*massfactor*frictionfactor*game::frametime.asSeconds()*(vel_y)/(sqrt(vel_x*vel_x + vel_y*vel_y));

}

bool gameObject::isInPocket()
{
    sf::Vector2f circlePos = circle.getPosition();
    if (circlePos.x + circlePos.y - board::pocket_offset < 0) {
        return 1;
    };
    if (circlePos.x + circlePos.y - 2*game::windowsize + board::pocket_offset > 0) {
        return 1;
    }
    if (circlePos.y - circlePos.x + game::windowsize - board::pocket_offset < 0) {
        return 1;
    }
    if (circlePos.y - circlePos.x - game::windowsize +  board::pocket_offset > 0) {
        return 1;
    }
    return 0;
}

void collision(gameObject *A, gameObject *B)
{
    sf::Vector2f posA = A->circle.getPosition();
    sf::Vector2f posB = B->circle.getPosition();
    float o = atan2(posB.y - posA.y, posB.x - posA.x);
    float ov = atan2(B->vel_y - A->vel_y, B->vel_x - A->vel_x);
    float sinx = sin(o);
    float cosx = cos(o);
    float sinvx = sin(ov);
    float cosvx = cos(ov);
    float dist = sqrt(pow(posA.x - posB.x, 2) + pow(posA.y - posB.y, 2));
    float orv = cosvx*cosx + sinvx*sinx;
    if (dist < A->circle.getRadius() + B->circle.getRadius() + 0.1f) {
        if (orv < 0) {
            swapvelocities(A, B);
        }
        //dirty hack to ensure intersection of sprites does not occur
        float D =  A->circle.getRadius() + B->circle.getRadius() + 0.5 - dist;
        float xmove = D*fabs(cos(o))/(A->circle.getRadius() + B->circle.getRadius());
        float ymove = D*fabs(sin(o))/(A->circle.getRadius() + B->circle.getRadius());
        if (posA.x < posB.x) {
            A->circle.move(-xmove*A->circle.getRadius(), 0);
            B->circle.move(xmove*B->circle.getRadius(), 0);
        }
        else {
            A->circle.move(xmove*A->circle.getRadius(), 0);
            B->circle.move(-xmove*B->circle.getRadius(), 0);
        }
        if (posA.y < posB.y) {
            A->circle.move(0, -ymove*A->circle.getRadius());
            B->circle.move(0, ymove*B->circle.getRadius());
        }
        else {
            A->circle.move(0, ymove*A->circle.getRadius());
            B->circle.move(0, -ymove*B->circle.getRadius());
        }
        game::_gamestate = game::_motion;
    }
}

void swapvelocities(gameObject *A, gameObject *B)
{
    sf::Vector2f posA = A->circle.getPosition();
    sf::Vector2f posB = B->circle.getPosition();
    float impact = sqrt(pow(A->vel_x - B->vel_x, 2) +
			pow(A->vel_y - B->vel_y, 2)); 
    //direction of the vector joining the centres of A and B
    float o = atan2(posB.y - posA.y, posB.x - posA.x);
    //velocities in a rotated coordinate system with the x-axis in the direction of 'o'
    float va, vaT, vb, vbT;

    va = A->vel_y*sin(o) + A->vel_x*cos(o);
    vaT = A->vel_y*cos(o) - A->vel_x*sin(o);


    vb = B->vel_y*sin(o) + B->vel_x*cos(o);
    vbT = B->vel_y*cos(o) - B->vel_x*sin(o);

    float vcm = (A->massfactor*va + B->massfactor*vb)/(A->massfactor +
						       B->massfactor);
    float vcma = va - vcm;
    float vcmb = vb - vcm;

    va = vcm - vcma*A->restitution;
    vb = vcm - vcmb*B->restitution;


    A->vel_x = va*cos(o) - vaT*sin(o);
    A->vel_y = vaT*cos(o) + va*sin(o);

    B->vel_x = vb*cos(o) - vbT*sin(o);
    B->vel_y = vbT*cos(o) + vb*sin(o);

    if (!(A->frictionfactor < 0.001f || B->frictionfactor < 0.001f)) {
	game::playsound(impact);
    }
}

int board::top;
int board::bottom;
int board::left;
int board::right;
int board::playzonein;
int board::playzoneout;
int board::pocket_offset;
sf::Vector2f striker::position;
float board::frictioncoefficient = 160.f;
float board::edgerestitution = 0.9f;
float gameObject::vel_init(0.00001f);


src/gameobjectmanager.cpp
#include "gameobjectmanager.hpp"
#include <math.h>

objectlist::objectlist(int n)
{
    size = n;
    list = new gameObject* [n];
    isActive = new bool [n];
    for (int i = 0; i < n; i++) {
        isActive[i] = 1; //All objects are drawn by default
    }
}

objectlist::~objectlist()
{
    for (int i = 0; i < size; i++) {
        delete list[i];
    }
    delete[] list;
    delete[] isActive;
}


void objectlist::checkCollision(int i, int j)
{
    if (isActive[i] && isActive[j]) {
        collision(list[i], list[j]);
    }
}

void objectlist::checkAllCollisions()
{
    for (int i = 0; i < size - 2; i++) {
        for (int j = i+1; j < size - 1; j++) {
            if(isActive[i] && isActive[j]) {
                checkCollision(i, j);
            }
        }
    }
}

void objectlist::updateall()
{
    if (game::_gamestate != game::_new) {
        list[0]->update();
    }
    for (int i = 1; i < size; i++) {
        if(isActive[i]) {
            list[i]->update();
        }
    }
}

void objectlist::setActive(int i)
{
    isActive[i] = 1;
}

void objectlist::setInActive(int i)
{
    isActive[i] = 0;
}

bool objectlist::getState(int i) const
{
    return isActive[i];
}

void objectlist::arrangeCoins()
{
    float r = list[1]->circle.getRadius();
    list[1]->circle.setPosition(game::windowsize/2.0,
				game::windowsize/2.0);

    list[2]->circle.setPosition(game::windowsize/2.0,
				game::windowsize/2.0 - 2*r);
    list[11]->circle.setPosition(game::windowsize/2.0 + sqrt(3)*r,
				 game::windowsize/2.0 - r);
    list[3]->circle.setPosition(game::windowsize/2.0 +sqrt(3)*r,
				game::windowsize/2.0 + r);
    list[12]->circle.setPosition(game::windowsize/2.0,
				 game::windowsize/2.0 + 2*r);
    list[4]->circle.setPosition(game::windowsize/2.0 - sqrt(3)*r,
				game::windowsize/2.0 + r);
    list[13]->circle.setPosition(game::windowsize/2.0 - sqrt(3)*r,
				 game::windowsize/2.0 - r);

    list[5]->circle.setPosition(game::windowsize/2.0,
				game::windowsize/2.0 - 4*r);
    list[14]->circle.setPosition(game::windowsize/2.0 + sqrt(3)*r,
				 game::windowsize/2.0 - 3*r);
    list[6]->circle.setPosition(game::windowsize/2.0 + 2*sqrt(3)*r,
				game::windowsize/2.0 - 2*r);
    list[15]->circle.setPosition(game::windowsize/2.0 + 2*sqrt(3)*r,
				 game::windowsize/2.0);
    list[7]->circle.setPosition(game::windowsize/2.0 + 2*sqrt(3)*r,
				game::windowsize/2.0 + 2*r);

    list[16]->circle.setPosition(game::windowsize/2.0 + sqrt(3)*r,
				 game::windowsize/2.0 + 3*r);
    list[8]->circle.setPosition(game::windowsize/2.0, game::windowsize/2.0 + 4*r);
    list[17]->circle.setPosition(game::windowsize/2.0 - sqrt(3)*r,
				 game::windowsize/2.0 + 3*r);
    list[9]->circle.setPosition(game::windowsize/2.0 - 2*sqrt(3)*r,
				game::windowsize/2.0 + 2*r);
    list[18]->circle.setPosition(game::windowsize/2.0 - 2*sqrt(3)*r,
				 game::windowsize/2.0);
    list[10]->circle.setPosition(game::windowsize/2.0 - 2*sqrt(3)*r,
				 game::windowsize/2.0 - 2*r);
    list[19]->circle.setPosition(game::windowsize/2.0 - sqrt(3)*r,
				 game::windowsize/2.0 - 3*r);
}

void objectlist::rotateCoins(sf::Event &event)
{
    for (int i = 1; i < 20; i++) {
        sf::Vector2f position = list[i]->circle.getPosition();
        sf::Vector2f relpos = sf::Vector2f(position.x - game::windowsize/2.0, position.y - game::windowsize/2.0);
        float radiusvec = sqrt(relpos.x*relpos.x + relpos.y*relpos.y);
        float o = atan2(relpos.y, relpos.x);
        if (sf::Joystick::isConnected(game::player)) {
            if (fabs(sf::Joystick::getAxisPosition(game::player, sf::Joystick::X)) > 0.5f) {
                list[i]->circle.setPosition(radiusvec*cos(o - 0.01f) + game::windowsize/2.0,
					    radiusvec*sin(o - 0.01f) + game::windowsize/2.0);
            }
        }
        else if (event.type == sf::Event::MouseMoved) {
            list[i]->circle.setPosition(radiusvec*cos(o - 0.01f) + game::windowsize/2.0,
					radiusvec*sin(o - 0.01f) + game::windowsize/2.0);
        }
    }
}


bool objectlist::hasAllStopped()
{
    for (int i = 0; i < 20; i++) {
        if (!((fabs(list[i]->vel_x) < 5.f && fabs(list[i]->vel_y) < 5.f) || list[i]->isInPocket())) {
            return 0;
        }
    }
    return 1;
}

int objectlist::isboardover()
{
    bool w = 1, b = 1;
    for (int i = 2; i < 11; i++) {
	if (isActive[i] == 1) {
	    w = 0;
	    break;
	}
    }
    for (int i = 11; i < 20; i++) {
	if (isActive[i] == 1) {
	    b = 0;
	    break;
	}
    }
    if (!isActive[1] && w) return 1;
    else if (!isActive[1] && b) return 2;
    return 0;
}


includes/ingameobjects.hpp
#ifndef IN_GAME_OBJECTS_H
#define IN_GAME_OBJECTS_H

#include <SFML/Graphics.hpp>

class board
{
private:
    sf::Sprite brd;
public:
    static int top;
    static int bottom;
    static int left;
    static int right;
    static int playzoneout;
    static int playzonein;
    static int pocket_offset;
    static float frictioncoefficient;
    static float edgerestitution;
    void setTexture();
    void draw();
};

class gameObject
{
protected:
    sf::CircleShape circle;
    sf::Sprite coinsprite;
    float massfactor;
    float vel_x;
    float vel_y;
    float restitution;
    float frictionfactor;
public:
    virtual ~gameObject() {};
    virtual void update() = 0;
    virtual void reset() {};
    void boundtoboard();
    void move();
    friend void collision(gameObject *A, gameObject *B);
    friend void swapvelocities(gameObject *A, gameObject *B);
    static float vel_init;
    bool isInPocket();
    friend class objectlist;
    friend class striker;
    friend class coin;
};

class pointer : public gameObject
{
private:
    sf::Vector2f relMousePos;
public:
    pointer();
    void update();
    void input();
    void boundaroundstriker();
    void hitStriker(float dist);
};

class striker : public gameObject
{
private:
    sf::Vector2f relMousePos;
public:
    static sf::Vector2f position;
    striker();
    void update();
    void reset();
private:
    void input();
    void boundtoplay();
};

class coin : public gameObject
{
private:
    char type; //w-white, b-black, other is red
public:
    coin(char t = 'q');
    void update();
    void reset();
};

#endif


src/game.cpp
#include <SFML/Graphics.hpp>
#include <SFML/Audio.hpp>
#include "ingameobjects.hpp"
#include "game.hpp"
#include "gameobjectmanager.hpp"

void game::init()
{
    _gamestate = _init;

    player = 0;
    turn = 0;
    turnincr = 1;
    sf::ContextSettings settings;
    settings.antialiasingLevel = 4;
    window.create(sf::VideoMode(windowsize, windowsize, 32), "gnarrom",
            sf::Style::Default);
    window.setVerticalSyncEnabled(1);
    //window.setFramerateLimit(60);
    window.setMouseCursorVisible(0);
    brd.loadFromFile("../resources/board.png");
    brd.setSmooth(1);
    B.setTexture();

    cointex.loadFromFile("../resources/coin.png");
    cointex.setSmooth(1);
    strikertex.loadFromFile("../resources/striker.png");
    strikertex.setSmooth(1);

    soundbuf.loadFromFile("../resources/sound.ogg");
    sound.setBuffer(soundbuf);
    sound.setLoop(0);

    _gamestate = _new;

    frametime = sf::seconds(0);

    loop();
}

void game::loop()
{
    //We do not intend the objectslist to have a lifetime beyond the scope of this function
    objects = new objectlist(21);

    objects->list[0] = new striker();
    objects->list[1] = new coin();

    for (int i = 2; i < 11; i++) {
        objects->list[i] = new coin('w');
    }
    for (int i = 11; i < 20; i++) {
        objects->list[i] = new coin('b');
    }

    objects->list[20] = new pointer();
    objects->arrangeCoins();

    sf::Thread endturnthread(&endturn);

    //--------------------------------------------------------------------------
    sf::Clock deltaTime;
    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            changeState(event);
            if (event.type == sf::Event::Closed ||
                    sf::Keyboard::isKeyPressed(sf::Keyboard::Escape)) {
                window.close();
            }

            if (_gamestate == _new) {
                captureMouse();
                objects->rotateCoins(event);
            }

            if (_gamestate == _quit) {
                window.close();
            }
            if (sf::Keyboard::isKeyPressed(sf::Keyboard::RShift)) {
                capmouse = 0;
            }
            else if (sf::Keyboard::isKeyPressed(sf::Keyboard::Return)) {
                capmouse = 1;
            }
        }

        window.clear(brdcolor);
        B.draw();

        if (objects->isboardover() > 0) {
            _gamestate = _quit;
        }

        if (_gamestate == _hit) {
            objects->checkCollision(0, 20);
        }

        if (_gamestate == _motion) {
            objects->checkAllCollisions();
            endturnthread.launch();
        }

        for (int i = 0; i < 20; i++) {
            if (objects->list[i]->isInPocket()) {
                if (objects->getState(i) && i != 0) {
                    processPocket(i);
                }
                objects->setInActive(i);
            }
        }

        frametime = deltaTime.restart();
        objects->updateall();
        window.display();
    }

    delete objects;
}

void game::endturn()
{
    if (_gamestate == _motion) {
        if (objects->hasAllStopped()) {
            if (objects->list[0]->isInPocket()) {
                processFine();
            }

            sf::sleep(sf::milliseconds(100));

            coinsinpocket = 0;
            turn += turnincr;
            turnincr = 1;
            player = turn % player_count;
            objects->list[0]->reset();
        }
    }
}

void game::captureMouse()
{
    if (!capmouse) return;
    sf::Vector2i mousePos = sf::Mouse::getPosition(window);
    if (mousePos.x > board::right || mousePos.x < board::left || mousePos.y > board::bottom || mousePos.y < board::top) {
        sf::Mouse::setPosition(sf::Vector2i(300, 300), window);
    }
}

void game::changeState(sf::Event &event)
{
    //Check if any joystick is connected for given player
    if (sf::Joystick::isConnected(player)) {

        if (_gamestate != _motion && (sf::Joystick::isButtonPressed(player, 0) || sf::Joystick::isButtonPressed(player, 1) ||
                    sf::Joystick::isButtonPressed(player, 2) || sf::Joystick::isButtonPressed(player, 3))) {
            if (_gamestate == _new) {
                _gamestate = _striker;
            }
            else {
                _gamestate = _pointer;
            }
        }
        if (_gamestate == _pointer && !(sf::Joystick::isButtonPressed(player, 0) || sf::Joystick::isButtonPressed(player, 1) ||
                    sf::Joystick::isButtonPressed(player, 2) || sf::Joystick::isButtonPressed(player, 3))) {
            _gamestate = _hit;
        }
    }
    else {
        if(_gamestate != _motion && event.type == sf::Event::MouseButtonPressed){
            if (_gamestate == _new) {
                _gamestate = _striker;
            }
            else {
                _gamestate = _pointer;
            }
        }
        else if(_gamestate == _pointer && event.type == sf::Event::MouseButtonReleased){
            _gamestate = _hit;
        }
    }
}

void game::playsound(float impact)
{
    if (!(sound.getStatus() == sf::Sound::Playing)) {
        if (impact < 2000.f) {
            sound.setVolume(impact/20.f);
        }
        else {
            sound.setVolume(100.f);
        }
        sound.play();
    }
}

void game::setwindowsize(int winsize)
{
    windowsize = winsize;
}

void game::close()
{
    _gamestate = _quit;
    window.close();
}

void game::processPocket(int n)
{
    if (player == 0 && n > 0 && n < 11) {
        turnincr = 0;
        coinsinpocket++;
    }
    else if (player == 1 && (n == 1 || (n > 10 && n < 20))) {
        turnincr = 0;
        coinsinpocket++;
    }
}

void game::processFine()
{
    do {
        if (player == 0) {
            for (int i = 2; i < 11; i++) {
                if (objects->getState(i) == 0) {
                    objects->list[i]->reset();
                    objects->setActive(i);
                    coinsinpocket -= 1;
                    break;
                }
            }
        }
        else if (player == 1) {
            for (int i = 11; i < 20; i++) {
                if (objects->getState(i) == 0) {
                    objects->list[i]->reset();
                    objects->setActive(i);
                    coinsinpocket -= 1;
                    break;
                }
            }
        }
    } while (coinsinpocket > 0);
}

void game::setColor(int brdR, int brdG, int brdB,
        int wR, int wG, int wB,
        int bR, int bG, int bB,
        int qR, int qG, int qB,
        int sR, int sG, int sB,
        int piR, int piG, int piB,
        int poR, int poG, int poB)
{
}

sf::RenderWindow game::window;
game::gameStates game::_gamestate;
sf::Texture game::brd;
sf::Texture game::cointex;
sf::Texture game::strikertex;
sf::Time game::frametime = sf::seconds(0);
objectlist *game::objects;
board game::B;
int game::player_count = 2;
int game::player = 0;
int game::turn = 0;
int game::turnincr = 1;
int game::windowsize = 700;
int game::coinsinpocket = 0; //stores the number of coins pocketed in present turn;
sf::Color game::brdcolor(221, 218, 216);
sf::Color game::wcolor = sf::Color::White;
sf::Color game::bcolor(50, 50, 50);
sf::Color game::qcolor = sf::Color::Red;
sf::Color game::scolor = sf::Color::Green;
sf::Color game::pointercolorin = sf::Color::Green;
sf::Color game::pointercolorout = sf::Color::Black;
bool game::capmouse = 1;
sf::SoundBuffer game::soundbuf;
sf::Sound game::sound;
